src/stable_marriage.rs

```diff
@@ -1,96 +1,234 @@
+) -> (Vec<(W, Option<M>)>, Vec<M>) {
+    let mut women = women
+        .into_iter()
+        .map(|item| Woman {
+            data: item,
+            current_match: None,
+            current_is_perfect: false,
+        })
+        .collect::<Vec<Woman<P>>>();
+
+    let mut no_longer_engageables = Vec::new();
+    'outer_loop: for item in men {
+        let mut man = Man {
+            data: item,
+            prefs: vec![],
+        };
+
+        if let Some(woman) = men_matcher.find_perfect_match(&man.data, &mut women) {
+            woman.current_is_perfect = true;
+            woman.replace_match(man);
+            continue;
+        }
+        man.prefs = men_matcher.compute_preference_list(&man.data, &women, women_matcher);
+
+        // Loop while we still hold an engageable man
+        while let Some(i) = man.prefs.pop() {
+            let woman = &mut women[i];
+            if woman.prefers_to_current(women_matcher, &man.data) {
+                if let Some(rejected_man) = woman.replace_match(man) {
+                    man = rejected_man;
+                } else {
+                    // We no longer hold a man; fetch the next one
+                    continue 'outer_loop;
+                }
+            }
+        }
+        // `man` has no remaining women he wants to propose to
+        no_longer_engageables.push(man);
+    }
+
+    (
+        women
+            .into_iter()
+            .map(|x| (x.data, x.current_match.map(|man| man.data)))
+            .collect_vec(),
+        no_longer_engageables
+            .into_iter()
+            .map(|man| man.data)
+            .collect_vec(),
+    )
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
+    use itertools::{self, Itertools};
+
+    struct IndexMatcher(Vec<Vec<usize>>);
+
+    impl Matcher for IndexMatcher {
+        type Item = usize;
+        type Target = usize;
+
+        fn is_admissible(&self, x: &Self::Item, y: &Self::Target) -> bool {
+            self.0[*x].iter().position(|j| j == y).is_some()
+        }
+
+        fn cmp_3way(
+            &self,
+            from: &Self::Item,
+            left: &Self::Target,
+            right: &Self::Target,
+        ) -> Ordering {
+            let pos_left = self.0[*from].iter().position(|j| j == left);
+            let pos_right = self.0[*from].iter().position(|j| j == right);
+            pos_left.cmp(&pos_right)
+        }
+
+
+        fn compute_preference_list<Q>(
+            &self,
+            item: &Self::Item,
+            _targets: &Vec<Woman<Self>>,
+            _other_matcher: &Q,
+        ) -> Vec<usize>
+        where
+            Q: Matcher<Item = Self::Target, Target = Self::Item>,
+        {
+            // Cheat, because we know which targets will be used
+            self.0[*item].iter().cloned().rev().collect_vec()
+        }
+    }
+
+    fn numbers() -> impl Iterator<Item = usize> {
+        itertools::unfold(0, |x| {
+            let v = *x;
+            *x += 1;
+            Some(v)
```

Comment:
Hmm… Isn't this `(0..)`?

Reply:
You're right, I didn't know about ranges before three days ago ^^